File: dron_cinematico/src/lemniscata_helice_grafica/lemniscata_helice_grafica.cpp
/* 
Descripcion:
Este nodo ejecuta el controlador para esolver el problema de seguimiento utilizando un dron en gazebo. 
También tanto la trayectoria deseada como la trayectoria real de dron si publicadas en los tópicos "/tr_path" y "/path", respectivamente, para su visualización en Rviz.

En la función "velocity_controller" la trayectoria deseada es generada y las señales de control son calculadas. 
Para este ejemplo, la trayectoria deseada es una leminscata.

Primero, el servicio "/enable_motors" es llamado para mover el cuadrirrotor simulado. 
Las funciones de Despegue y Aterrizaje son definidas.

Las velocidades traslacionales (Vx, Vy, Vz) y la velocidad rotacional (Wz) son publicadas en el tópico "/cmd_vel".
La posición (x,y,z) y la orientación (quaternion) son recibidas desde el tópico "/ground_truth/state".



INSTRUCCIONES
Para aterrizar el dron y finalizar este nodo, presiones la tecla 'ESC'.

En diferentes terminales:
$ roslaunch my_hector_uavs my_hector_uav.launch
	# Lanza un cuatrirrotor (my simplified drone) en Gazebo usando un mundo vacío. 
	  Nota: La simulacióm emoieza pausada.

$ rosrun my_hector_uavs hector_quadrotor_controller_paths
	# Corre este nodo
	
$ roslaunch my_hector_uavs visualization_paths.launch
	# Corre Rviz con mi archivo de configuración para visualizar el dron y las trajectorias

*/
#include "ros/ros.h"
#include "hector_uav_msgs/EnableMotors.h" 	// Hacer la llamada al servicio --> roservice info/enable_motors
#include "nav_msgs/Odometry.h"				// Para suscribirle a la geometria del dron (mensaje)
#include "geometry_msgs/Twist.h"			// Para publicar velocidades (mensaje)
#include "tf/transform_broadcaster.h"		// Para hacer la conversion de cuaternion a angulo-Euler (fichero)
#include "nav_msgs/Path.h" 					// Para visualizar trayectorias en Rviz
// Los ultimos 3 se necesitan para utilizar eventos del teclado (función kbhit implemetada en Linux).
#include <termios.h>
#include <unistd.h>
#include <fcntl.h>

using namespace std;

// Define objetos globales
ros::Publisher  velocity_publisher;
ros::Subscriber pose_subscriber;
geometry_msgs::Twist vel;
ros::ServiceClient enable_motors_client;

double x, y, z, roll, pitch, yaw; // Defino variables globales (posicion y orientacion, usando ángulos de Euler)
float takeoff_alt = 1.2; // Altura inicial deseada
int key;


geometry_msgs::PoseStamped drone_pose, tr_pose; // Crear objetos globales para definir la posición del dron y la trayectoria deseada usando mensajes de tipo geometry_msgs/Twist

// Define objetos para publicar la trayectoria del dron y la trayectoria deseada en Rviz
ros::Publisher drone_path_pub, tr_path_pub;
nav_msgs::Path drone_path_msg, tr_path_msg; // Objetos trayectoria de tipo nav_msgs/Path


// Crear el título del marco global. El marco global es requerido por Rviz para definir el MISMO marco para cualquier robot
const string global_frame = "/world";

double t, t0, Vxy_max = 1, Vz_max = 1.0, Wz_max = 4; /// Temporizador, tiempo inicial, velocidades traslacionales máximas [m/s] y máxima velocidad rotacional [rad/s]

double T = 100, k = 0.1; // Período de tayectoria, ganancias de controlador kx = ky = k
double ex, ey, ez, e_yaw, Xd, Yd, Zd, Yawd, Xdp, Ydp, Zdp, Yawdp; // Errores de seguimiento, posiciones deseadas, orientación deseada y sus derivadas con respecto al tiempo, respectivamente

int kbhit(void){ // Funcion para utilizar eventos de teclado
  struct termios oldt, newt;
  int ch, oldf;

  tcgetattr(STDIN_FILENO, &oldt);
  newt = oldt;
  newt.c_lflag &= ~(ICANON | ECHO);
  tcsetattr(STDIN_FILENO, TCSANOW, &newt);
  oldf = fcntl(STDIN_FILENO, F_GETFL, 0);
  fcntl(STDIN_FILENO, F_SETFL, oldf | O_NONBLOCK);

  ch = getchar();

  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
  fcntl(STDIN_FILENO, F_SETFL, oldf);

  if(ch != EOF)
  {
    ungetc(ch, stdin);
    return 1;
  }
  return 0;
}

void enable_motors(){ // Función para llamar al servicio "/ enable_motors", necesaria para mover el dron
  ROS_WARN_ONCE("Calling the enable_motors service...\n");
  hector_uav_msgs::EnableMotors enable_request;
  enable_request.request.enable = true;
  enable_motors_client.call(enable_request);

  if(enable_request.response.success){ ROS_WARN_ONCE("Enable motors successful\n"); }
  else{ cout << "Enable motors failed" << endl; }
}

void poseCallback(const nav_msgs::Odometry::ConstPtr msg){ // Función de devolución de llamada para obtener la posición del dron
  x = msg->pose.pose.position.x;
  y = msg->pose.pose.position.y;
  z = msg->pose.pose.position.z;
  // Operaciones para convertir de cuaternión a ángulos de Euler
  tf::Quaternion quater;
  tf::quaternionMsgToTF(msg->pose.pose.orientation, quater);
  tf::Matrix3x3(quater).getRPY(roll, pitch, yaw);  // Los ángulos de Euler están dados en Radianes

  drone_pose.pose.position = msg->pose.pose.position; // Copia la posición el dron al objeto "drone_pose"
  drone_path_msg.poses.push_back(drone_pose); // Agregar la nueva posición a la lista
}

void movement(float x, float y, float z, float turn){ // Función para asignar señales de control (Vx, Vy, Vz, Wz)
  vel.linear.x = x;       vel.linear.y = y;
  vel.linear.z = z;       vel.angular.z = turn;
}

void takeoff(){ //Takeoff function
  movement(0,0,0.2,0);
  velocity_publisher.publish(vel);

  cout << " Desired takeoff altitude = " << takeoff_alt << "\n Taking off ...\n";
  while(z < takeoff_alt-0.1){ // Comparar si se ha logrado la altura deseada
    velocity_publisher.publish(vel); 
    ros::spinOnce(); ros::Duration(0.1).sleep();
    if(kbhit())	key = getchar();
    if(key == 27) break;
  }
  movement(0,0,0,0);
  velocity_publisher.publish(vel);
}

void land(){ // Función de Aterrizaje
  movement(0,0,-0.2,0);
  velocity_publisher.publish(vel);

  cout << "\n Landing ...\n";
  while(z > 0.3){
    velocity_publisher.publish(vel);
    ros::spinOnce(); ros::Duration(0.1).sleep();
  }
  movement(0,0,0,0);
  velocity_publisher.publish(vel);
}

void velocity_controller(){ // Función para generar la trayectoria deseada y para calcular las señales de control
  double Vx, Vy, Vz, Wz; // Define las señales de control signals del drone

  // Trayectoria deseada: Lemniscata
  double a = 3, b = 1.5, X0 = 0, Y0 = -0.5, Z0 = 1.5, w = 2*M_PI/T, c = 0.5, d = M_PI/2;
  // Posición deseada en el espacio 3D
  Xd = X0+a*sin(w*t);
  Yd = Y0+b*sin(2*w*t);
  Zd = Z0+c*sin(w*t); //Nota: 1 <= Zd <= 2
  Yawd = d*sin(w*t);

  // Derivadas con respecto al tiempo
  Xdp = a*w*cos(w*t);
  Ydp = 2*b*w*cos(2*w*t);
  Zdp = c*w*cos(w*t);
  Yawdp = d*w*cos(w*t);

  // Asignar la posición deseada al oobjeto "tr_pose"
  tr_pose.pose.position.x = Xd; tr_pose.pose.position.y = Yd; tr_pose.pose.position.z = Zd;
  tr_path_msg.poses.push_back(tr_pose); // Agregar el nuevo punto a la lista


  ex = x-Xd; ey = y-Yd; ez = z-Zd; e_yaw = yaw - Yawd; // Calcular errores de seguimiento

  // Controlador Cinemático. Controles Auxiliares, en coordenadas globales
  double Ux = Xdp-k*ex; double Uy = Ydp-k*ey;
  
  // Velocidades traslacionales con respecto al marco del robot
  Vx = Ux*cos(yaw)+Uy*sin(yaw);
  Vy = -Ux*sin(yaw)+Uy*cos(yaw);

  // Controlador Cinemático. Nota: Vz y Wzson calculadas directamente,ya que ellas son dadas con respecto al marco del robot
  Vz = Zdp-k*ez;
  Wz = Yawdp-k*e_yaw;

  // Velocidades de saturación
  if(abs(Vx)>Vxy_max){Vx = Vxy_max*abs(Vx)/Vx; printf("Sat Vx\t");}
  if(abs(Vy)>Vxy_max){Vy = Vxy_max*abs(Vy)/Vy; printf("Sat Vy\t");}
  if(abs(Vz)>Vz_max){Vz = Vz_max*abs(Vz)/Vz; printf("Sat Vz\t");}
  if(abs(Wz)>Wz_max){Wz = Wz_max*abs(Wz)/Wz; printf("Sat Wz\t");}

  movement(Vx,Vy,Vz,Wz);
  velocity_publisher.publish(vel); // Publicar las 4 señales de control
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "hector_quadrotor_controller_paths");  // Inicializar el nodo
  ros::NodeHandle n; // Crear el nodo Handle
  int freq = 50; // Asignar frecuencia (Hz) de ejecucion al nodo
  ros::Rate loop_rate(freq);
  int counter = 0;
  bool finish = false;

  enable_motors_client = n.serviceClient<hector_uav_msgs::EnableMotors>("enable_motors"); // Se configura  y llama al servicio
  velocity_publisher = n.advertise<geometry_msgs::Twist>("/cmd_vel", 1); // Publica en el topico velocidades utilizando el mensaje geometry 
  pose_subscriber = n.subscribe("/ground_truth/state", 1, poseCallback); //Se suscribe al topico de odometria usando el topico ground y se asigna una funcion de llamada de vuelta 

  // Publicar las trayectorias
  drone_path_pub = n.advertise<nav_msgs::Path>("/path", 10);
  tr_path_pub = n.advertise<nav_msgs::Path>("/tr_path", 10);

  //Importante: Asignación del MIMSMO marco de referencia para TODOS los robots
  drone_path_msg.header.frame_id = tr_path_msg.header.frame_id = global_frame;

  enable_motors(); //Ejecuta las funciones "enable_motors" y "takeoff"
  takeoff();
 
  printf("Press 'ESC' to land the drone and finish the node");
  printf("To clean up both paths, press 'c' key\n");
  ROS_WARN_ONCE("To start the movement, the simulation must be running\n\n"); // Mensaje de advertencia

  t0 = ros::Time::now().toSec(); // Obtener el tiempo inical  

  do{
    t = ros::Time::now().toSec()-t0; //printf("t: %.1f\n",t); // Calculo el tiempo del controlador     
    velocity_controller(); // Calculo las señales de control

    if(counter == 50){ // Divisior de frecuencia para mostrar los errores de seguimiento de forma mas lenta 
      printf("ex: %.3f ey: %.3f ez: %.2f e_yaw: %.3f\n", ex,ey,ez,e_yaw); // Imprimier en la terminal los errores de seguimiento [m] y el error de orientacion [rad]
      counter = 0; //Resetear el contador
    }else counter++;

    // Publica los mensajes "path"
    drone_path_pub.publish(drone_path_msg); tr_path_pub.publish(tr_path_msg);

    ros::spinOnce();    //Requerido para recibir funciones de devolución de llamada (call back)
    loop_rate.sleep();   //Comando para esperar el resto del tiempo para completar la frecuencia del bucle
    if(kbhit())	key = getchar();

    if(key == 'c'){ // Limpiar ambas trayectorias
      drone_path_msg.poses.clear(); tr_path_msg.poses.clear(); // Limpiar mensajes de trayectoria
      ROS_WARN_ONCE("Clear paths\n");
      key = 0; //Resetear la tecla
    }
    if(key == 27) finish = true;
  }while(finish == false); //Note: No se utiliza ros::ok() porque es necesario seguir utilizando funciones de ros
  
  land(); // Ejecuta la función Aterrizar dron
  printf("\n Node finished\n");
  return 0;
}

File: dron_cinematico/scripts/lemniscata_helice_grafica/lemniscata_helice_grafica.py
#!/usr/bin/env python

#Use 'python' for ROS Kinetic and Melodic.
#Use 'python3' for ROS Noetic
"""This python script is the equivalent to the 'src/hector_quadrotor_controller_paths.cpp', so it executes the same tasks.
For more information, please refers to that source file.

-- Important: To land the drone and finish this node, please press 'q' key."""

__author__ = "C. Mauricio Arteaga-Escamilla"

import rospy, math
from hector_uav_msgs.srv import EnableMotors #rosservice info /enable_motors 
from nav_msgs.msg import Odometry, Path #Message type to subscribe to /odom and to publish in /path topics
from geometry_msgs.msg import Twist
from tf.transformations import euler_from_quaternion, quaternion_from_euler
from math import sin, cos, pi
from geometry_msgs.msg import PoseStamped #Message type to publish the robots' path

#kbhit function implemented on Linux
import sys, select, os #Handling command line arguments
if os.name == 'nt':
  import msvcrt
else:
  import tty, termios

#Define global variables (position and orientation, using Euler anlges)
x=0; y=0; z=0; roll=0; pitch=0; yaw=0
takeoff_alt = 1.2 #Desired initial altitude

vel = Twist() #Create Twist and Path message instances
drone_path_msg = Path()
tr_path_msg = Path()

#Create the title of the global frame. Global frame is required by Rviz to define the SAME frame for any robot
global_frame = "/world"

t = 0.0; t0 = 0.0; Vxy_max = 1.5;  Vz_max = 0.5; Wz_max = 4; #timer, initial time, maximum velocities [m/s, rad/s], respectively
T = 100.0; k = 0.1; #Trajectory period, controller gains kx = ky = k
ex = 0.0; ey = 0.0; ez = 0; e_yaw = 0

def getKey(): #Function to use keyboard events on Linux
    if os.name == 'nt':
      return msvcrt.getch()

    tty.setraw(sys.stdin.fileno())
    rlist, _, _ = select.select([sys.stdin], [], [], 0.1)
    if rlist:
        key = sys.stdin.read(1)
    else:
        key = ''

    termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
    return key

def enable_motors(): #Function to call "/enable_motors" service, needed to move the drone
    SERVICE_ENABLE_MOTORS = "enable_motors"
    print("Waiting for service", SERVICE_ENABLE_MOTORS)
    rospy.wait_for_service(SERVICE_ENABLE_MOTORS)
    try:
        enable_motors = rospy.ServiceProxy(SERVICE_ENABLE_MOTORS, EnableMotors)
        res = enable_motors(True)
        if res:
            print("Enable motors successful\n")
        else:
            print("Enable motors failed\n")
    except rospy.ServiceException:
        print("Enable service", SERVICE_ENABLE_MOTORS, "call failed")

def poseCallback(msg): #Callback function to get the drone posture
        global x, y, z, roll, pitch, yaw
        x = msg.pose.pose.position.x
        y = msg.pose.pose.position.y
        z = msg.pose.pose.position.z
        #Operations to convert from quaternion to Euler angles
        quater = msg.pose.pose.orientation
        quater_list = [quater.x, quater.y, quater.z, quater.w]
        (roll, pitch, yaw) = euler_from_quaternion(quater_list)
        
        drone_pose = PoseStamped() #Create the PoseStamped instance
        drone_pose.pose.position = msg.pose.pose.position
        drone_path_msg.poses.append(drone_pose) #Add the drone position to the list of points

def movement(x,y,z,turn): #Function to assign control signals (Vx, Vy, Vz, Wz)
	vel.linear.x = x; vel.linear.y = y
	vel.linear.z = z; vel.angular.z = turn

def takeoff(): #Takeoff function
	movement(0,0,0.5,0)
	velocity_publisher.publish(vel)
	
	rospy.logwarn(" Desired takeoff altitude = %.2f\n Taking off ...\n", takeoff_alt)
	while(z < takeoff_alt-0.1):
		velocity_publisher.publish(vel)
		rate.sleep(); rospy.sleep(0.1)
		key = getKey()
		if(key == b'\x1b'):
			break
	
	movement(0,0,0,0)
	velocity_publisher.publish(vel)

def land(): #Land function
	movement(0,0,-0.2,0)
	velocity_publisher.publish(vel)
	print("\n Landing ...\n")
	
	while(z > 0.3):
		velocity_publisher.publish(vel)
		rate.sleep(); rospy.sleep(0.1)
	
	movement(0,0,0,0)
	velocity_publisher.publish(vel)

def velocity_controller(): #Function to generate the desired trajectory and to compute the control signals
	global ex, ey, ez, e_yaw #Indicate that some variables are global to be used in the main_function-----
	#Desired trajectory: Lemniscate
	a = 3; b = 1.5; X0 = 0; Y0 = -0.5; Z0 = 1.5; w = 2*pi/T; c = 0.5; d = pi/2
	#Desired position in the 3D space 
	Xd = X0+a*sin(w*t)
	Yd = Y0+b*sin(2*w*t)
	Zd = Z0+c*sin(w*t) #Note: 1 <= Zd <= 2
	Yawd = d*sin(w*t)
	
	#Corresponding time derivatives
	Xdp = a*w*cos(w*t)
	Ydp = 2*b*w*cos(2*w*t)
	Zdp = c*w*cos(w*t)
	Yawdp = d*w*cos(w*t)
	
	#Corresponding time derivatives
	Xdp = a*w*cos(w*t)
	Ydp = 2*b*w*cos(2*w*t)
	Zdp = c*w*cos(w*t)
	Yawdp = d*w*cos(w*t)
	
	
	#Assign the desired position to "tr_pose"
	tr_pose = PoseStamped() #Create the PoseStamped instance
	tr_pose.pose.position.x = Xd; tr_pose.pose.position.y = Yd
	tr_pose.pose.position.z = Zd 
	tr_path_msg.poses.append(tr_pose) #Add the new point to the list
	
	
	ex = x-Xd; ey = y-Yd; ez = z-Zd; e_yaw = yaw-Yawd #Compute tracking errors
	
	#Cinematic controller. Auxiliar controls, in global coordinates
	Ux = Xdp-k*ex; Uy = Ydp-k*ey
	
	#Translational velocities with respect to the robot frame
	Vx = Ux*cos(yaw)+Uy*sin(yaw)
	Vy = -Ux*sin(yaw)+Uy*cos(yaw)
	
	#Cinematic controller. Note: Vz and Wz are compute directly since they are given with respect to the robot frame
	Vz = Zdp-k*ez
	Wz = Yawdp-k*e_yaw

	#Velocities saturation
	if(abs(Vx)>Vxy_max):
		Vx = Vxy_max*abs(Vx)/Vx; print("Sat Vx\t")
	if(abs(Vy)>Vxy_max):
		Vy = Vxy_max*abs(Vy)/Vy; print("Sat Vy\t")
	if(abs(Vz)>Vz_max):
		Vz = Vz_max*abs(Vz)/Vz; print("Sat Vz\t")
	if(abs(Wz)>Wz_max):
		Wz = Wz_max*abs(Wz)/Wz; print("Sat Wz\t")
	
	movement(Vx,Vy,Vz,Wz)
	velocity_publisher.publish(vel) #Publish the 4 control signals


def main_function():
	rospy.init_node("hector_quadrotor_controller_paths", anonymous=True) #Initialize the node. anonymous=True for multiple nodes
	global rate
	rate = rospy.Rate(50) #Node frequency (Hz)
	counter = 0
	
	global velocity_publisher
	velocity_publisher = rospy.Publisher('/cmd_vel', Twist, queue_size=10) #To publish in the topic
	rospy.Subscriber('/ground_truth/state',Odometry, poseCallback) #To subscribe to the topic
	
	#To publish the paths
	drone_path_pub = rospy.Publisher('/path', Path, queue_size=10)
	tr_path_pub = rospy.Publisher('/tr_path', Path, queue_size=10) 
	
	global drone_path_msg, tr_path_msg
	#Important: Assignation of the SAME reference frame for ALL robots
	drone_path_msg.header.frame_id = tr_path_msg.header.frame_id = global_frame

	enable_motors() #Execute the functions
	takeoff()
	
	print("Press 'q' to land the drone and finish the node\n")
	print("To clean up both paths, press 'c' key\n")
	rospy.logwarn("To start the movement, the simulation must be running\n\n") #Warning message
	
	global t, t0 #t and t0 are global to be used in velocity_controller()
	t0 = rospy.Time.now().to_sec()
	
	while(1):
		t = rospy.Time.now().to_sec()-t0
		velocity_controller() #Compute the control signals
		
		if counter == 50: #Frequency divisor
			rospy.loginfo("ex: %.3f ey: %.3f ez: %.2f e_yaw: %.3f\n", ex,ey,ez,e_yaw) #Print tracking errors [m] and orientation error [rad]
			counter = 0 #Reset the counter
		else:
			counter += 1
		
		#Publish the "path" messages
		drone_path_pub.publish(drone_path_msg); tr_path_pub.publish(tr_path_msg)
		
		rate.sleep() #spinOnce() function does not exist in python
		key = getKey()
		if(key == 'c'): #Clear paths lists
			del drone_path_msg.poses[:]; del tr_path_msg.poses[:]
			rospy.logwarn("Clear paths\n")
			key = 0; #Reset the key
		elif(key == 'q'):
			break

	land() #Execute land function
	print("\n Node finished\n")


if __name__ == '__main__':
    if os.name != 'nt':
        settings = termios.tcgetattr(sys.stdin)

    try:
    	main_function()  #Execute the function
	
    except rospy.ROSInterruptException:
        pass

    #if os.name != 'nt':
    #   termios.tcsetattr(sys.stdin, termios.TCSADRAIN, settings)
